### PR TITLE
Fix GTP path echo timer reset

### PR DESCRIPTION
**Pull request by fholzhauser**
This PR would fix the issue that short sessions on a right side path (e.g. with GSN path change on create session response) resets the scheduled echo and can cause lack of path management in case only such sessions are done on a path.

- the gtp path sends an echo immediately when the path starts, speeding up the detection of a broken destination
- the echoes are scheduled relative to the last sent echo which prevents that path management stops due to rescheduling the echos just with the interval at each change

### CODE REVIEW

File: src/gtp_path.erl [hunk 1]
```diff
@@ -563,7 +576,7 @@ send_echo_request(State, #{socket := Socket, handler := Handler, ip := DstIP,
     Ref = erlang:make_ref(),
     CbInfo = {?MODULE, handle_response, [self(), echo_request, Ref]},
     ergw_gtp_c_socket:send_request(Socket, any, DstIP, ?GTP1c_PORT, T3, N3, Msg, CbInfo),
-    State#state{echo = Ref}.
+    State#state{echo = {Ref, os:system_time(millisecond)}}.
```
Comment (RoadRunnr) [inline src/gtp_path.erl:579]: this should use `erlang:monotic_time`

File: src/gtp_path.erl [hunk 2]
```diff
@@ -403,17 +406,27 @@ enter_state_timeout_action(down, #{down_timeout := Timeout}) when is_integer(Tim
...
-enter_state_echo_action(_, _) ->
+    {{timeout, echo}, next_echo_schedule(EchoInterval, LastEchoTs), start_echo};
+enter_state_echo_action(down, {_, LastEchoTs}, #{down_echo := EchoInterval})
+  when is_integer(EchoInterval) ->
+    {{timeout, echo}, next_echo_schedule(EchoInterval, LastEchoTs), start_echo};
+enter_state_echo_action(_, _, _) ->
     {{timeout, stop_echo}, 0, stop_echo}.
 
+next_echo_schedule(EchoInterval, LastEchoTs) ->
+    TS = os:system_time(millisecond),
```
Comment (RoadRunnr) [inline src/gtp_path.erl:422]: this should use `erlang:monotic_time`

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
I'm not sure that this change goes into the right direction.
The intention of the Echo logic was always to restart Echo interval when a state change occurs. The reasoning is that a change of state somehow indicates a activity on the peer and that that defines the last know state of the peer. The next Echo will then validate that state (or change it).

The real problem that this PR tries to fix is that pending requests and context will impact the state of the peer. And that is IMHO the real problem, activity triggered by us should never change the peers state, only responses and timeouts from the peer should impact the peer state.

File: src/gtp_path.erl (hunk 1 above)
Comment (fholzhauser) [reply src/gtp_path.erl]: yes, thanks for spotting it

File: src/gtp_path.erl (hunk 2 above)
Comment (fholzhauser) [reply src/gtp_path.erl]: indeed

**Comment by fholzhauser** [comment]
> I'm not sure that this change goes into the right direction.
> The intention of the Echo logic was always to restart Echo interval when a state change occurs. The reasoning is that a change of state somehow indicates a activity on the peer and that that defines the last know state of the peer. The next Echo will then validate that state (or change it).
> 
> The real problem that this PR tries to fix is that pending requests and context will impact the state of the peer. And that is IMHO the real problem, activity triggered by us should never change the peers state, only responses and timeouts from the peer should impact the peer state.

Yes indeed, the intention of this is not to change the existing logic, which is at the moment mixing echo probe results and contexts to determine the state, and as you mentioned could (and probably should) be fundamentally reconsidered. It only intends to fix a particular situation that can render broken path detection completely dysfunctional in a specific scenario. I'm also not quite sure if it changes the path state logic, but only the scheduling of the echo probes to make sure that the interval changes at state transitions don't result in echoes not being sent.

**Comment by mgumz** [comment]
@fholzhauser is this PR address https://github.com/travelping/smc/issues/37 ?

**Comment by RoadRunnr** [comment]
> @fholzhauser is this PR address [travelping/smc#37](https://github.com/travelping/smc/issues/37) ?

It might, but it is the wrong approach (IMHO).

The underlying problem is that outgoing GTP requests are considered as a livesign of the peer and therefore reset the Echo timer. The correct approach would be to only use GTP packets from the peer (requests and responses) as indication of liveliness.

This PR does not change that. Instead it decouples the GTP echo timer from the other activity indications.

**Comment by RoadRunnr** [comment]
for the 3.x and master branch this is resolved in #428

Is this PR till needed or can we close it?

**Comment by vkatsuba** [comment]
PR https://github.com/travelping/ergw/pull/428 is related to master(`3.X`) version. Current PR is related to `2.X`. Depending on how critical the problem for `2.X` - we can close this PR or provide release `2.8.15` with including fix. However I not sure that this is should be done. I will close current PR and  if this will be necessary to fix it for `2.8.X` too, we always can back to current implementation.

**Pull request closed without merging**